### ISSUE TITLE
JVM can't protect protection page

### ISSUE THREAD

**Issue by cpriebe** [lsds/sgx-lkl#17]
I am adding this issue for documentation purposes.

Recently, running more complex applications on the JVM on top of SGX-LKL fails with the following JVM error:

```shell
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (virtualspace.cpp:325), pid=1, tid=0x00007fef70080a40
#  fatal error: cannot protect protection page
#
# JRE version:  (8.0_121-b13) (build )
# Java VM: OpenJDK 64-Bit Server VM (25.121-b13 mixed mode linux-amd64 compressed oops)
# Derivative: IcedTea 3.3.0
# Distribution: Custom build (Thu Jul  6 09:24:47 UTC 2017)
```

The JVM fails to protect the protection page via `mprotect([...], PROT_NONE)`. SGX-LKL forwards `mprotect` calls to the host as SGX version 1 currently does not support changes to the SGX-specific EPC page permissions, so we have to rely on the host OS. The corresponding `mprotect` call fails with an `EACCES` error. However, the JVM protection page lies within the enclave region and is mapped at the time of the `mprotect` call so that it should succeed.

**Comment by cpriebe** [comment]
This is most likely due to a Linux kernel bug introduced by the changes related to the L1TF/Foreshadow mitigations (see https://lkml.org/lkml/2018/8/14/887).

The actual bug and the fix are described here: https://lkml.org/lkml/2018/8/16/697

```[...]
Manifested as an unexpected mprotect(..., PROT_NONE) failure when
called on a VMA that has VM_PFNMAP and was mmap'd to as something
other than PROT_NONE but never used.  mprotect() sends the PROT_NONE
request down prot_none_walk(), which walks the PTEs to check the PFNs.
prot_none_pte_entry() gets the bogus PFN from pte_pfn() and returns
-EACCES because it thinks mprotect() is trying to adjust a high MMIO
address```

This has already been patched upstream. Looking at the Ubuntu kernel changelogs (https://launchpad.net/ubuntu/+source/linux/+changelog, search for "x86/speculation/l1tf: Exempt zeroed PTEs from inversion"), it seems like the patch has been applied to the 4.17 kernel used by Ubuntu 18.10, but hasn't made it into the 16.04/18.04 LTS kernels.

To summarize, the solution here is to use a recent version of the kernel that includes the `x86/speculation/l1tf: Exempt zeroed PTEs from inversion` patch.

**Comment by cpriebe** [comment]
This seems to have been fixed in 14.04/16.04/18.04, so I am closing this issue.

**Comment by cpriebe** [comment]
This has been resolved a while ago in the official 16.04 and 18.10 Ubuntu kernels. This finally has been backported to the 4.15 kernel used by Ubuntu 18.04. The new kernel version that includes the fix has officially been released today.

See the corresponding bug report here: https://bugs.launchpad.net/ubuntu/+bug/1799237

For Ubuntu 18.04 the bug is fixed in kernel version 4.15.0-46.49.

I am closing this issue.

**Comment by Zxiaoyu95** [comment]
hi @cpriebe ，Recently,I meet the same error when i  running  applications on the JVM on top of SGX-LKL(ubuntu16.04  4.15.0-46-generic ),to resolve the probleam,i must update ubuntu 16.04 to ubuntu 18.04?

**Comment by cpriebe** [comment]
Hi @Zxiaoyu95 

I am not entirely sure whether the fix is available as part of the Ubuntu 16.04 4.15 image (4.15.0-46.49~16.04.1). But it should be fixed in the 4.4 kernel that is used by 16.04 by default. You could downgrade to that. As you suggested, you could also update to 18.04. In any case, you need to use a kernel with the fix applied. Upstream this has been fixed from I believe 4.17.17 and onward.

You can check which of the official Ubuntu kernel images have the fix applied to them here:

https://launchpad.net/ubuntu/+source/linux/+changelog

Search for `x86/speculation/l1tf: Exempt zeroed PTEs from inversion` to find where and when the fix has been backported.

**Comment by Zxiaoyu95** [comment]
Ok,thank you @cpriebe ,now after  update my ubuntu to 18.04,this problem has been resolved.
As the changelog shows,the 4.4.150 kernel and 4.17 kernel also resolved this bug.i believe it is available!
Thank you!
